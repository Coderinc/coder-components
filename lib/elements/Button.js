import React from 'react';
import styled from 'styled-components';

const ButtonDiv = styled.div`
  background-color: ${props => {
    if (props.disabled) {return props.theme.offsetBackground;}

    if (props.secondary || props.ternary) {return props.theme.activeBackground;}
    return props.theme[props.color || 'primary'];
  }};

  color: ${props => {
    if (props.disabled) {return props.theme.offsetFont;}

    if (props.ternary) {return props.theme.offsetFont;}
    else if (props.secondary) {return props.theme[props.color || 'primary'];}
    return props.theme.primaryFont;
  }};

  margin: 1rem 0.5rem;
  padding: 0.75rem;
  font-size: 0.875rem;
  letter-spacing: 0.0625rem;
  outline: none;
  border: none;
  cursor: ${props => (props.disabled ? 'auto' : 'pointer')};
  border-radius: 0.25rem;
  text-transform: uppercase;
  font-weight: 200;
  width: ${props => props.width}px;
  max-width: 100%;
  position: relative;
  text-align: ${props => (props.reversed ? 'right' : 'left')};
  transition: 0.5s ease all;

  box-shadow: ${props =>
    props.disabled
      ? undefined
      : '0 1px 3px rgba(0, 0, 0, 0.02), 0 1px 2px rgba(0, 0, 0, 0.04)'};

  &::before,
  &::after {
    position: absolute;
    content: '\f100';
    font-family: coder-icons;
    font-size: 1rem;
    transition: 0.5s ease all;
  }

  &::before {
    transform: scaleX(-1);
    display: ${props => (props.reversed ? undefined : 'none')};
    left: 24px;
  }

  &:hover::before {
    left: ${props => (props.disabled ? undefined : '16px')};
  }

  &::after {
    display: ${props => (props.reversed ? 'none' : undefined)};
    right: 24px;
  }

  &:hover::after {
    right: ${props => (props.disabled ? undefined : '16px')};
  }

  &:hover {
    color: ${props =>
    props.ternary ? props.theme[props.color || 'primary'] : undefined};
    box-shadow: ${props =>
    props.disabled
      ? undefined
      : '0 5px 10px rgba(0, 0, 0, 0.25), 0 5px 5px rgba(0, 0, 0, 0.10)'};
  }
`;

export const Button = props => <ButtonDiv {...props}>{props.label}</ButtonDiv>;

export default Button;

