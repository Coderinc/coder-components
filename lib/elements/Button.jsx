import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const ButtonDiv = styled.div`
  background-color: ${({
    disabled, secondary, ternary, theme, color,
  }) => {
    if (disabled) {
      return theme.offsetBackground;
    }

    if (secondary || ternary) {
      return theme.activeBackground;
    }
    return theme[color || 'primary'];
  }};

  color: ${({
    disabled, secondary, ternary, theme, color,
  }) => {
    if (disabled) {
      return theme.offsetFont;
    }

    if (ternary) {
      return theme.offsetFont;
    }
    if (secondary) {
      return theme[color || 'primary'];
    }
    return theme.primaryFont;
  }};

  margin: 1rem 0.5rem;
  padding: 0.75rem;
  font-size: 0.875rem;
  letter-spacing: 0.0625rem;
  outline: none;
  border: none;
  cursor: ${({ disabled }) => (disabled ? 'auto' : 'pointer')};
  border-radius: 0.25rem;
  text-transform: uppercase;
  font-weight: 200;
  width: ${({ width }) => width}px;
  max-width: 100%;
  position: relative;
  text-align: ${({ reversed }) => (reversed ? 'right' : 'left')};
  transition: 0.5s ease all;

  box-shadow: ${({ disabled }) => (disabled ? undefined : '0 1px 3px rgba(0, 0, 0, 0.02), 0 1px 2px rgba(0, 0, 0, 0.04)')};

  &::before,
  &::after {
    position: absolute;
    content: '\f100';
    font-family: coder-icons;
    font-size: 1rem;
    transition: 0.5s ease all;
  }

  &::before {
    transform: scaleX(-1);
    display: ${({ reversed }) => (reversed ? undefined : 'none')};
    left: 24px;
  }

  &:hover::before {
    left: ${({ disabled }) => (disabled ? undefined : '16px')};
  }

  &::after {
    display: ${({ reversed }) => (reversed ? 'none' : undefined)};
    right: 24px;
  }

  &:hover::after {
    right: ${({ disabled }) => (disabled ? undefined : '16px')};
  }

  &:hover {
    color: ${({ ternary, theme, color }) => (ternary ? theme[color || 'primary'] : undefined)};
    box-shadow: ${({ disabled }) => (disabled ? undefined : '0 5px 10px rgba(0, 0, 0, 0.25), 0 5px 5px rgba(0, 0, 0, 0.10)')};
  }
`;

const Button = ({ label, ...props }) => (
  <ButtonDiv {...props}>
    {label}
  </ButtonDiv>
);

Button.propTypes = {
  label: PropTypes.string,
};

Button.defaultProps = {
  label: 'Click',
};

export default Button;
