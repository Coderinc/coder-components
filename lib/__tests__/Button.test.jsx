import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import sinon from 'sinon';
import 'jest-styled-components';
import Button from '../elements/Button';

describe('Button', () => {
  it('renders correctly', () => {
    const component = renderer.create(<Button />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('sets color correctly', () => {
    const component = renderer.create(<Button color="blue" theme={{ blue: '#0000FF' }} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('swaps button styles correctly', () => {
    const component = renderer.create(<Button color="blue" secondary theme={{ blue: '#0000FF' }} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('fire onClick', () => {
    const onClick = sinon.spy();
    const component = mount(<Button onClick={onClick} />);
    component.find('div').simulate('click');
    expect(onClick.calledOnce).toEqual(true);
  });
});
