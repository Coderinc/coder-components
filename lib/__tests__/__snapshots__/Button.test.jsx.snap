// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
.c0 {
  margin: 1rem 0.5rem;
  padding: 0.75rem;
  font-size: 0.875rem;
  -webkit-letter-spacing: 0.0625rem;
  -moz-letter-spacing: 0.0625rem;
  -ms-letter-spacing: 0.0625rem;
  letter-spacing: 0.0625rem;
  outline: none;
  border: none;
  cursor: pointer;
  border-radius: 0.25rem;
  text-transform: uppercase;
  font-weight: 200;
  width: px;
  max-width: 100%;
  position: relative;
  text-align: left;
  -webkit-transition: 0.5s ease all;
  transition: 0.5s ease all;
  box-shadow: 0 1px 3px rgba(0,0,0,0.02),0 1px 2px rgba(0,0,0,0.04);
}

.c0::before,
.c0::after {
  position: absolute;
  content: '\\f100';
  font-family: coder-icons;
  font-size: 1rem;
  -webkit-transition: 0.5s ease all;
  transition: 0.5s ease all;
}

.c0::before {
  -webkit-transform: scaleX(-1);
  -ms-transform: scaleX(-1);
  transform: scaleX(-1);
  display: none;
  left: 24px;
}

.c0:hover::before {
  left: 16px;
}

.c0::after {
  right: 24px;
}

.c0:hover::after {
  right: 16px;
}

.c0:hover {
  box-shadow: 0 5px 10px rgba(0,0,0,0.25),0 5px 5px rgba(0,0,0,0.10);
}

<div
  className="c0"
  color="primary"
>
  Click
</div>
`;

exports[`Button sets color correctly 1`] = `
.c0 {
  background-color: #0000FF;
  margin: 1rem 0.5rem;
  padding: 0.75rem;
  font-size: 0.875rem;
  -webkit-letter-spacing: 0.0625rem;
  -moz-letter-spacing: 0.0625rem;
  -ms-letter-spacing: 0.0625rem;
  letter-spacing: 0.0625rem;
  outline: none;
  border: none;
  cursor: pointer;
  border-radius: 0.25rem;
  text-transform: uppercase;
  font-weight: 200;
  width: px;
  max-width: 100%;
  position: relative;
  text-align: left;
  -webkit-transition: 0.5s ease all;
  transition: 0.5s ease all;
  box-shadow: 0 1px 3px rgba(0,0,0,0.02),0 1px 2px rgba(0,0,0,0.04);
}

.c0::before,
.c0::after {
  position: absolute;
  content: '\\f100';
  font-family: coder-icons;
  font-size: 1rem;
  -webkit-transition: 0.5s ease all;
  transition: 0.5s ease all;
}

.c0::before {
  -webkit-transform: scaleX(-1);
  -ms-transform: scaleX(-1);
  transform: scaleX(-1);
  display: none;
  left: 24px;
}

.c0:hover::before {
  left: 16px;
}

.c0::after {
  right: 24px;
}

.c0:hover::after {
  right: 16px;
}

.c0:hover {
  box-shadow: 0 5px 10px rgba(0,0,0,0.25),0 5px 5px rgba(0,0,0,0.10);
}

<div
  className="c0"
  color="blue"
>
  Click
</div>
`;

exports[`Button swaps button styles correctly 1`] = `
.c0 {
  color: #0000FF;
  margin: 1rem 0.5rem;
  padding: 0.75rem;
  font-size: 0.875rem;
  -webkit-letter-spacing: 0.0625rem;
  -moz-letter-spacing: 0.0625rem;
  -ms-letter-spacing: 0.0625rem;
  letter-spacing: 0.0625rem;
  outline: none;
  border: none;
  cursor: pointer;
  border-radius: 0.25rem;
  text-transform: uppercase;
  font-weight: 200;
  width: px;
  max-width: 100%;
  position: relative;
  text-align: left;
  -webkit-transition: 0.5s ease all;
  transition: 0.5s ease all;
  box-shadow: 0 1px 3px rgba(0,0,0,0.02),0 1px 2px rgba(0,0,0,0.04);
}

.c0::before,
.c0::after {
  position: absolute;
  content: '\\f100';
  font-family: coder-icons;
  font-size: 1rem;
  -webkit-transition: 0.5s ease all;
  transition: 0.5s ease all;
}

.c0::before {
  -webkit-transform: scaleX(-1);
  -ms-transform: scaleX(-1);
  transform: scaleX(-1);
  display: none;
  left: 24px;
}

.c0:hover::before {
  left: 16px;
}

.c0::after {
  right: 24px;
}

.c0:hover::after {
  right: 16px;
}

.c0:hover {
  box-shadow: 0 5px 10px rgba(0,0,0,0.25),0 5px 5px rgba(0,0,0,0.10);
}

<div
  className="c0"
  color="blue"
>
  Click
</div>
`;
